### PR TITLE
[Kraken] private channel

### PR DESCRIPTION
**Pull request by mdvx**
Performace issue fixes
1) allow getOrderChanges/getUserTrades now always create a single channel subscription
2) a better nonce factory

### CODE REVIEW

File: xchange-kraken/src/main/java/info/bitrich/xchangestream/kraken/KrakenStreamingTradeService.java [hunk 1]
```diff
@@ -136,22 +140,25 @@ else if ("market".equals(orderType))
 
     @Override
     public Observable<UserTrade> getUserTrades(CurrencyPair currencyPair, Object... args) {
-            try {
+        try {
```
Comment (badgerwithagun) [inline xchange-kraken/src/main/java/info/bitrich/xchangestream/kraken/KrakenStreamingTradeService.java]: Same comments apply here.

File: xchange-kraken/src/main/java/info/bitrich/xchangestream/kraken/KrakenStreamingTradeService.java [hunk 2]
```diff
@@ -54,18 +56,20 @@ private KrakenWebsocketToken renewToken() throws IOException {
...
+                        .filter(JsonNode::isArray)
+                        .filter(Objects::nonNull)
+                        .map(jsonNode -> jsonNode.get(0))
+                        .map(jsonNode ->
+                                StreamingObjectMapperHelper.getObjectMapper().treeToValue(jsonNode, KrakenDtoOrderHolder[].class))
+                        .flatMapIterable(this::adaptKrakenOrders);
+            }
+            return ownTradesObservable
+                    .filter(order -> currencyPair == null || order.getCurrencyPair() == null || order.getCurrencyPair().compareTo(currencyPair) == 0)
+                    .share();
```
[the remark this answers is not in the document]
Comment (mdvx) [reply xchange-kraken/src/main/java/info/bitrich/xchangestream/kraken/KrakenStreamingTradeService.java]: does this work better?
```
public Observable<Order> getOrderChanges(CurrencyPair currencyPair, Object... args) {
        try {
            if (ownTradesObservable == null) {
                String channelName = getChannelName(KrakenSubscriptionName.openOrders);
                ownTradesObservable = streamingService.subscribeChannel(channelName, renewToken().getToken())
                        .filter(JsonNode::isArray)
                        .filter(Objects::nonNull)
                        .map(jsonNode -> jsonNode.get(0))
                        .map(jsonNode ->
                                StreamingObjectMapperHelper.getObjectMapper().treeToValue(jsonNode, KrakenDtoOrderHolder[].class))
                        .flatMapIterable(this::adaptKrakenOrders)
                        .share();

                return ownTradesObservable;
            }
            return Observable.create( emit -> ownTradesObservable
                        .filter(order -> currencyPair == null || order.getCurrencyPair() == null || order.getCurrencyPair().compareTo(currencyPair) == 0)
                        .subscribe(emit::onNext, emit::onError, emit::onComplete));

        } catch (IOException e) {
            return Observable.error(e);
        }
    }
```

[the remark this answers is not in the document]
Comment (badgerwithagun) [reply xchange-kraken/src/main/java/info/bitrich/xchangestream/kraken/KrakenStreamingTradeService.java]: Adding another `if` outside the `synchronized` block, as in the example I gave below, is better. Double-check locking is a cool trick (worth googling).

[the remark this answers is not in the document]
Comment (badgerwithagun) [reply xchange-kraken/src/main/java/info/bitrich/xchangestream/kraken/KrakenStreamingTradeService.java]: Oh, also, that doesn't work - you need to remove the `return ownTradesObservable`.

[the remark this answers is not in the document]
Comment (mdvx) [reply xchange-kraken/src/main/java/info/bitrich/xchangestream/kraken/KrakenStreamingTradeService.java]: That should do it

**Pull request merged**